torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -651,20 +634,21 @@ Message tensorpipeDeserialize(
     tensors.emplace_back(std::move(t));
   }
 
-  if (!buffers.deviceIndices.empty()) {
-    TORCH_INTERNAL_ASSERT(
-        buffers.deviceIndices.size() == tensors.size(),
-        "Number of device indices must match the number of tensors in the "
-        "RPC message. But got ",
-        tensors.size(),
-        " tensors with ",
-        buffers.deviceIndices.size(),
-        " device indices.");
-    for (size_t i = 0; i < tensors.size(); ++i) {
-      auto index = buffers.deviceIndices[i];
-      if (tensors[i].device().index() != index) {
-        tensors[i] = tensors[i].to(indexToDevice(index));
-      }
+  // NB: This is a temporary solution. When TensorPipe Tensor.data can point to
+  // a CUDA memory address, we should directly use CUDACachingAllocator to
+  // create CUDA buffers in tensorpipeAllocate.
+  for (size_t i = 0; i < message.tensors.size(); ++i) {
+    auto& tensor = message.tensors[i];
+    if (!tensor.metadata.empty()) {
+      TORCH_INTERNAL_ASSERT(
```

Comment:
Is there a reason to have this check inside the for loop? It doesn't seem to depend on the value of `i`, does it?

Reply:
This check actually won't hold if tensors in RPC args or return values share storage. The pickler might pickle the storage instead the real tensors, and as a result, the tensor number in TpMessage and RpcMessage can be different. So I put a temporary check here to only verify the equivalence when there are non-empty metadata (i.e., newly introduced tests).  Will remove this when we can pass CUDA DataPtr to TP.